tests/ui/impl-trait/in-trait/refine-resolution-errors.rs

```diff
@@ -0,0 +1,25 @@
+// This is a non-regression test for issue #126670 where RPITIT refinement checking encountered
+// errors during resolution and ICEd.
+
+//@ edition: 2018
+
+pub trait A {}
+
+pub trait Mirror {
+    type Assoc: ?Sized;
+}
+impl<T: ?Sized> Mirror for dyn A {
+    //~^ ERROR the type parameter `T` is not constrained
+    type Assoc = T;
+}
+
+pub trait First {
+    async fn first() -> <dyn A + 'static as Mirror>::Assoc;
+    //~^ ERROR type annotations needed
+}
+
+impl First for dyn A {
+    async fn first() {}
+}
+
+fn main() {}
```

Comment:
This can be simplified. It really doesn't have much to do with `dyn Trait`s:

```
pub trait Mirror {
    type Assoc;
}
impl<T: ?Sized> Mirror for () {
    //~^ ERROR the type parameter `T` is not constrained
    type Assoc = T;
}

pub trait First {
    async fn first() -> <() as Mirror>::Assoc;
    //~^ ERROR type annotations needed
}

impl First for () {
    async fn first() {}
}

fn main() {}
```

Reply:
Good catch yeah. Done